medcat-v2/medcat/components/linking/embedding_linker.py

```diff
@@ -0,0 +1,583 @@
+        """Infer all entities at once (or in batches), to avoid multiple gpu calls
+        when it isn't nessescary.
+        Args:
+            doc (BaseDocument): The document look in.
+            entities (list[BaseEntity]): The entities to infer.
+        Yields:
+            entity (MutableEntity): Entity with a relevant cui prediction -
+            or skip if it's not suitable."""
+        detected_context_vectors = self._get_context_vectors(
+            doc, entities, self.cnf_l.context_window_size
+        )
+
+        # score all detected contexts vs all names
+        names_scores = detected_context_vectors @ self.names_context_matrix.T
+        cui_scores = detected_context_vectors @ self.cui_context_matrix.T
+        sorted_indices = torch.argsort(names_scores, dim=1, descending=True)
+
+        for i, entity in enumerate(entities):
+            link_candidates = [
+                cui
+                for cui in entity.link_candidates
+                if self.cnf_l.filters.check_filters(cui)
+            ]
+            if len(link_candidates) == 1:
+                best_idx = self._cui_to_idx[link_candidates[0]]
+                predicted_cui = link_candidates[0]
+                similarity = names_scores[i, best_idx].item()
+            elif len(link_candidates) > 1:
+                name_to_cuis = defaultdict(list)
+                for cui in link_candidates:
+                    for name in self.cdb.cui2info[cui]["names"]:
+                        name_to_cuis[name].append(cui)
+
+                name_idxs = [self._name_to_idx[name] for name in name_to_cuis]
+                indexed_scores = names_scores[i, name_idxs]
+
+                best_local_pos = int(torch.argmax(indexed_scores).item())
+                best_global_idx = name_idxs[best_local_pos]
+                similarity = names_scores[i, best_global_idx].item()
+                best_name = self._name_keys[best_global_idx]
+                cuis = name_to_cuis[best_name]
+                if len(cuis) == 1:
+                    predicted_cui = cuis[0]
+                else:
+                    predicted_cui, _ = self._disambiguate_by_cui(cuis, cui_scores[i, :])
+            else:
+                row_sorted = sorted_indices[i]  # sorted candidate indices for entity i
+
+                # Find the first candidate in this row with CUIs
+                first_true_pos = int(
+                    torch.nonzero(self._valid_names[row_sorted], as_tuple=True)[0][
+                        0
+                    ].item()
+                )
+
+                # Get global index + name
+                top_name_idx = int(row_sorted[first_true_pos].item())
+                similarity = names_scores[i, top_name_idx].item()
+                detected_name = self._name_keys[top_name_idx]
+                cuis = list(self.cdb.name2info[detected_name]["per_cui_status"].keys())
+
+                predicted_cui, _ = self._disambiguate_by_cui(cuis, cui_scores[i, :])
+
```

Comment:
There should be a check for the filters here - if the "best match" isn't allowed, don't use it.

See comment from above for more detail.